projects/storefrontlib/src/cms-components/configurator/generic/configure-product/configure-product.component.ts

```diff
@@ -1,50 +1,37 @@
-import {
-  ChangeDetectionStrategy,
-  ChangeDetectorRef,
-  Component,
-  Input,
-  OnDestroy,
-  OnInit,
-} from '@angular/core';
-import { GenericConfigurator, Product } from '@spartacus/core';
-import { Observable, Subscription } from 'rxjs';
-import { filter } from 'rxjs/operators';
+import { ChangeDetectionStrategy, Component, Optional } from '@angular/core';
+import { GenericConfigurator } from '@spartacus/core';
+import { map } from 'rxjs/operators';
+import { OutletContextData } from '../../../../cms-structure/outlet/outlet.model';
 import { CurrentProductService } from '../../../product/current-product.service';
+//import { ProductContext } from '../../../product/product-outlets.model';
 
 @Component({
   selector: 'cx-configure-product',
   templateUrl: './configure-product.component.html',
   changeDetection: ChangeDetectionStrategy.OnPush,
 })
-export class ConfigureProductComponent implements OnInit, OnDestroy {
-  @Input() productCode: string;
-  @Input() configuratorType: string;
-
-  product$: Observable<Product> = this.currentProductService.getProduct();
-  subscription: Subscription;
+export class ConfigureProductComponent {
+  productContext: OutletContextData;
   ownerTypeProduct: GenericConfigurator.OwnerType =
     GenericConfigurator.OwnerType.PRODUCT;
 
   constructor(
-    private currentProductService: CurrentProductService,
-    private changeDetector: ChangeDetectorRef
-  ) {}
-
-  ngOnInit(): void {
-    if (!this.productCode) {
-      this.subscription = this.currentProductService
-        .getProduct()
-        .pipe(filter(Boolean))
-        .subscribe((product: Product) => {
-          this.productCode = product.code;
-          this.configuratorType = product.configuratorType;
-          this.changeDetector.markForCheck();
-        });
-    }
-  }
-  ngOnDestroy() {
-    if (this.subscription) {
-      this.subscription.unsubscribe();
+    protected currentProductService: CurrentProductService,
+    @Optional() protected outletContext?: OutletContextData
+  ) {
+    if (outletContext) {
+      this.productContext = outletContext;
+    } else {
+      //in this case component was instantiated via CMS, and no
+      //outlet context is available
+      this.productContext = {
+        reference: undefined,
+        position: undefined,
+        context: undefined,
+        context$: this.currentProductService
+          .getProduct()
+          .pipe(map((prod) => ({ product: prod }))),
+      };
```

Comment:
Generally, it would be good if child components don't know they are projected inside outlet (I mean ideally we should avoid direct on `OutletContextData`, but rather hide it under some _facade/proxy_). It will help in the future to be flexible and replace the `OutletContextData` with some other source of data, when we decide so.

For `CartItemComponent`, we've introduced `CartItemContext`.

Here I would not introduce special `ProductContext` token, but rather leverage the existing `CurrentProductService`. it fits our needs - it exposes data of the current product (_current_ - in terms of current context).

The implementation of `CurrentProductService` (consumed mainly in PDP) reads the current product code from the context of the current route and returns the resolved product data.

To use `CurrentProductService` inside Product List/Grid Components, we only need to provide locally this service, but using use the locally given _code_.

My idea is to:
1. `{ provide: CurrentProductService, useClass: LocalCurrentProductService }` locally in `@Component` decorator's `providers: []` section of `ProductGridItem` and `ProductListItem`
2. in `ngOnChanges` of both components, call `localCurrentProductService.setCode(product.code)`

Here is the example implementation of `LocalCurrentProductService`:
Then in our custom `LocalCurrentProductService` we can do:
```typescript
// I suggets it having as private (don't expose in public API)
@Injectable()
export class LocalCurrentProductService extends CurrentProductService {
  protected readonly code$ = new BehaviorSubject<string>(undefined);

  setCode(code: string) {
    this.code$.next(code);
  }

  /** @override */
  protected getCode() {
    return this.code$.pipe(distinctUntilChanged());
  }
}
```

Please note that original `CurrentProductService` doesn't have method `getCode()` yet, which we could extend in the child class. But it can be eaisly added without breaking change.


**Thanks to that, in the lower-level components (like `configure-product.component`) we can seamlessly use the `CurrentProductService`** 🎉